### PR TITLE
Add query results cache support for label names and label values APIs

### LINKED ISSUE #5395: Proposal: cache label names/values API response in the query results cache
## Problem

We see some customers repetitively sending the same requests (with the same parameters) over and over to the `GET /api/v1/label/<label_name>/values` endpoint ([doc](https://prometheus.io/docs/prometheus/latest/querying/api/#querying-label-values)).

Such requests can be heavy, in particular when the `start` and `end` timerange are missing because by default Prometheus queries the whole time range (since the beginning of the universe until the end of it).

Mimir supports the `-store.max-labels-query-length` limit, which limits the time range (end - start time) of series, label names and values queries. The default is 0, which means "no limit". Moreover this limit is currently enforced only when querying the store-gateways, but not ingesters. At Grafana Labs we set this limit to 768h (32 days).

## Proposal

Similarly to https://github.com/grafana/mimir/pull/5212, I propose to add the support for a (short-lived) query results cache for the label values API endpoint.

### Cache key

The "label values" API parameters are ([doc](https://prometheus.io/docs/prometheus/latest/querying/api/#querying-label-values)):
- `start`: Start timestamp. Optional.
- `end`: End timestamp. Optional.
- `match[]`: Repeated series selector argument that selects the series from which to read the label values. Optional.

I propose to add all parameters to the cache key.

Both TSDB (so the Mimir ingester) and Mimir store-gateway query the label values out of blocks overlapping within the `start` and `end` time. This means that for maximum granularity is the block. Since the smallest blocks we have are 2h blocks, I propose to align the `start` and `end` time to 2h boundaries when computing the cache key. This means that, if we receive two requests with different start/end time but within the same 2h range (aligned to block boundaries) the cache key for the two requests is the same (assuming the matchers are the same as well).

### Cache TTL

Similarly to https://github.com/grafana/mimir/pull/5212, I propose to add a new per-tenant configuration option to set the caching TTL for "label query".

### Extension to label names API endpoint

The same exact approach can be adopted for the "label names" API endpoint ([docs](https://prometheus.io/docs/prometheus/latest/querying/api/#getting-label-names)).

### PR DESCRIPTION
**Pull request by pracucci**
#### What this PR does
Similarly to https://github.com/grafana/mimir/pull/5212, in this PR I'm adding the support to cache label names and label values responses. See details at: https://github.com/grafana/mimir/issues/5395

Note to reviewers:
- I manually tested it and looks working as intended.
- I opted for a new configuration option to specify the labels query response cache TTL, instead of one in common with the cardinality API, because I think we can potentially cache a bit longer for cardinality APIs. Since these config options are still marked experimental (so we're not committing to them yet) I prefer to keep them separated while rolling out the feature.

#### Which issue(s) this PR fixes or relates to

Fixes #5395

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/frontend/querymiddleware/roundtrip.go
Comment (pracucci) [inline pkg/frontend/querymiddleware/roundtrip.go:32]: Note to reviwers: adding the `/api/v1/` prefix is not a breaking change because these are the prefixes mapped in `pkg/api/api.go`.

File: pkg/frontend/querymiddleware/labels_query_cache.go
Comment (dimitarvdimitrov) [inline pkg/frontend/querymiddleware/labels_query_cache.go:33]: These look taken from here https://github.com/grafana/mimir/blob/989fd4d523c84324f9bd8f0072a10e52b06cc2e3/integration/util.go#L45-L51

And they're also used in `time_test.go`. WDYT about moving them to [time.go](https://github.com/grafana/mimir/blob/6030ca6446e0782d1e9e2adf4cb39a3cb9b26418/pkg/util/time.go) and exporting them?

File: pkg/frontend/querymiddleware/labels_query_cache.go
Comment (dimitarvdimitrov) [inline pkg/frontend/querymiddleware/labels_query_cache.go:129]: nitpick, but do we need the count?

File: pkg/frontend/querymiddleware/labels_query_cache.go
Comment (dimitarvdimitrov) [inline pkg/frontend/querymiddleware/labels_query_cache.go:141]: There is `util.MultiMatchersStringer` which does the same as what you implemented in this function. Can we reuse it?

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM with a few questions

File: pkg/frontend/querymiddleware/labels_query_cache.go
Comment (pracucci) [reply pkg/frontend/querymiddleware/labels_query_cache.go]: Good idea. Done here: https://github.com/grafana/mimir/pull/5426/commits/c3844b9e88a5c169439f58e54b49fa3229969b28

Comment (pracucci) [reply pkg/frontend/querymiddleware/labels_query_cache.go]: Also opened a PR to expose them (if Prometheus maintainers are open to it) https://github.com/prometheus/prometheus/pull/12535 but I will not block on it.

File: pkg/frontend/querymiddleware/labels_query_cache.go
Comment (pracucci) [reply pkg/frontend/querymiddleware/labels_query_cache.go]: I wasn't aware of that util. Thanks for pointing out. Adopted here: https://github.com/grafana/mimir/pull/5426/commits/f228944de4b79805a20d058016447981463c6d13

**Pull request merged**
